File: go/tokens.go
package agenttrust

import (
	"context"
	"net/http"
)

// TokensAPI provides opaque agent token issuance, introspection, and revocation.
//
// Tokens are random opaque strings prefixed "at_" — not JWTs. They have no
// signature and cannot be validated client-side. To check a token, call
// (*TokensAPI).Introspect.
type TokensAPI struct {
	client *Client
}

// Issue creates a new opaque agent token (prefix "at_").
//
// The returned Token contains the opaque string in Token.Token to use in
// Authorization: Bearer headers. Tokens are short-lived (default 300 seconds)
// and scoped to specific permissions.
//
// Example:
//
//	token, err := client.Tokens.Issue(ctx, ati.IssueTokenRequest{
//	    AgentID:  agent.ID,
//	    Scope:    []string{"files:read"},
//	    TTL:      300,
//	})
//	// Use token.Token in requests:
//	//   Authorization: Bearer <token.Token>
func (t *TokensAPI) Issue(ctx context.Context, req IssueTokenRequest) (*Token, error) {
	if req.TTL == 0 {
		req.TTL = 300
	}
	if req.Scope == nil {
		req.Scope = []string{}
	}
	var resp issueTokenResponse
	if err := t.client.doRequest(ctx, http.MethodPost, "/api/v1/agent-tokens/issue", req, &resp); err != nil {
		return nil, err
	}
	return parseTokenFromResponse(&resp), nil
}

// Introspect sends a token to POST /api/v1/agent-tokens/introspect for
// server-side validation. Returns {active, agent_id, org_id, scopes,
// expires_at}.
//
// Example:
//
//	result, err := client.Tokens.Introspect(ctx, ati.IntrospectTokenRequest{
//	    Token:          token.Token,
//	    Target:         "mcp://filesystem",
//	    RequiredScopes: []string{"files:read"},
//	})
//	if result.Active {
//	    // Allow the operation
//	}
func (t *TokensAPI) Introspect(ctx context.Context, req IntrospectTokenRequest) (*IntrospectionResult, error) {
	if req.RequiredScopes == nil {
		req.RequiredScopes = []string{}
	}

	var resp IntrospectionResult
	if err := t.client.doRequest(ctx, http.MethodPost, "/api/v1/agent-tokens/introspect", req, &resp); err != nil {
		return nil, err
	}
	if resp.Scopes == nil {
		resp.Scopes = []string{}
	}
	return &resp, nil
}

// Verify is a deprecated alias for Introspect.
//
// Deprecated: Use (*TokensAPI).Introspect.
func (t *TokensAPI) Verify(ctx context.Context, req VerifyTokenRequest) (*VerificationResult, error) {
	return t.Introspect(ctx, req)
}

// Revoke immediately invalidates a specific opaque agent token.
func (t *TokensAPI) Revoke(ctx context.Context, token string, reason string) error {
	if reason == "" {
		reason = "manual_revocation"
	}
	return t.client.doRequest(ctx, http.MethodPost, "/api/v1/agent-tokens/revoke", revokeTokenRequest{
		Token:  token,
		Reason: reason,
	}, nil)
}

File: rust/src/tokens.rs
//! Opaque agent token issuance, introspection, and revocation.
//!
//! Tokens are random opaque strings prefixed `at_` — not JWTs. They have no
//! signature and cannot be validated client-side. To check a token, call
//! [`TokensAPI::introspect`].
//!
//! # Example
//!
//! ```rust,no_run
//! use agenttrustid::{AgentTrustClient, IssueTokenRequest, IntrospectTokenRequest};
//!
//! let client = AgentTrustClient::builder()
//!     .base_url("http://localhost:8080")
//!     .build()
//!     .unwrap();
//!
//! // Issue a token
//! let token = client.tokens().issue(&IssueTokenRequest {
//!     agent_id: "agent-123".to_string(),
//!     scope: vec!["files:read".to_string()],
//!     audience: vec!["mcp://filesystem".to_string()],
//!     ttl: 300,
//! }).unwrap();
//!
//! // Use the opaque token in Authorization headers
//! println!("Bearer {}", token.token);
//!
//! // Introspect a token
//! let result = client.tokens().introspect(&IntrospectTokenRequest {
//!     token: token.token.clone(),
//!     target: Some("mcp://filesystem".to_string()),
//!     required_scopes: vec!["files:read".to_string()],
//! }).unwrap();
//!
//! if result.active {
//!     println!("Access granted");
//! }
//! ```

use crate::client::AgentTrustClient;
use crate::error::Result;
use crate::models::{
    IntrospectTokenRequest, IntrospectionResult, IssueTokenRequest, IssueTokenResponse,
    RevokeTokenRequest, Token,
};

/// Provides opaque agent token issuance, introspection, and revocation.
///
/// Obtained via [`AgentTrustClient::tokens()`].
pub struct TokensAPI<'a> {
    pub(crate) client: &'a AgentTrustClient,
}

impl<'a> TokensAPI<'a> {
    /// Issue a new opaque agent token (prefix `at_`) for an agent.
    ///
    /// The returned [`Token`] contains the opaque string in `Token.token` to
    /// use in `Authorization: Bearer` headers. Tokens are short-lived
    /// (default 300 seconds) and scoped to specific permissions.
    ///
    /// # Errors
    ///
    /// - [`AgentTrustError::Validation`](crate::AgentTrustError::Validation) if required fields are missing.
    /// - [`AgentTrustError::NotFound`](crate::AgentTrustError::NotFound) if the agent does not exist.
    pub fn issue(&self, req: &IssueTokenRequest) -> Result<Token> {
        let resp: IssueTokenResponse =
            self.client
                .request("POST", "/api/v1/agent-tokens/issue", Some(req))?;
        Ok(resp.into_token())
    }

    /// Introspect a token by calling `POST /api/v1/agent-tokens/introspect`.
    ///
    /// Returns `{ active, agent_id, org_id, scopes, expires_at, ... }`. The
    /// platform performs all validation server-side; opaque tokens have no
    /// signature so client-side checks aren't possible.
    ///
    /// # Errors
    ///
    /// Returns an error if the request itself fails (network, auth, etc.).
    /// A successful return with `active: false` is not an error.
    pub fn introspect(&self, req: &IntrospectTokenRequest) -> Result<IntrospectionResult> {
        let resp: IntrospectionResult =
            self.client
                .request("POST", "/api/v1/agent-tokens/introspect", Some(req))?;
        Ok(resp)
    }

    /// Deprecated alias for [`TokensAPI::introspect`].
    #[deprecated(note = "Use TokensAPI::introspect")]
    pub fn verify(&self, req: &IntrospectTokenRequest) -> Result<IntrospectionResult> {
        self.introspect(req)
    }

    /// Immediately invalidate a specific opaque agent token.
    ///
    /// # Arguments
    ///
    /// * `token` - The opaque token string (prefix `at_`) to revoke.
    /// * `reason` - Human-readable reason for the revocation.
    pub fn revoke(&self, token: &str, reason: &str) -> Result<()> {
        let body = RevokeTokenRequest {
            token: token.to_string(),
            reason: reason.to_string(),
        };
        self.client
            .request_no_response("POST", "/api/v1/agent-tokens/revoke", Some(&body))
    }
}
